File: goblinframework-embedded/src/main/java/org/goblinframework/embedded/core/provider/JavaEmbeddedServerFactory.java
package org.goblinframework.embedded.core.provider;

import org.goblinframework.embedded.core.EmbeddedServer;
import org.goblinframework.embedded.core.EmbeddedServerFactory;
import org.goblinframework.embedded.core.EmbeddedServerMode;
import org.goblinframework.embedded.core.manager.EmbeddedServerSetting;
import org.jetbrains.annotations.NotNull;

public class JavaEmbeddedServerFactory implements EmbeddedServerFactory {

  @NotNull
  @Override
  public EmbeddedServerMode mode() {
    return EmbeddedServerMode.JAVA;
  }

  @NotNull
  @Override
  public EmbeddedServer createEmbeddedServer(@NotNull EmbeddedServerSetting setting) {
    return new JavaEmbeddedServer(setting);
  }
}

File: goblin-embedded/src/main/java/org/goblinframework/embedded/java/JavaEmbeddedServerFactory.kt
package org.goblinframework.embedded.java

import org.goblinframework.api.annotation.Singleton
import org.goblinframework.embedded.core.setting.ServerSetting
import org.goblinframework.embedded.server.EmbeddedServer
import org.goblinframework.embedded.server.EmbeddedServerFactory
import org.goblinframework.embedded.server.EmbeddedServerMode

@Singleton
class JavaEmbeddedServerFactory private constructor() : EmbeddedServerFactory {

  companion object {
    @JvmField val INSTANCE = JavaEmbeddedServerFactory()
  }

  override fun mode(): EmbeddedServerMode {
    return EmbeddedServerMode.JAVA
  }

  override fun createEmbeddedServer(setting: ServerSetting): EmbeddedServer {
    if (setting.mode() !== mode()) {
      throw UnsupportedOperationException()
    }
    return JavaEmbeddedServer(setting)
  }
}